### PR TITLE
[3006.x] Fix user.list_groups omits remote groups

### PR DESCRIPTION
**Pull request by nicholasmhughes**
### What does this PR do?
See issues for details.

### What issues does this PR fix or reference?
Fixes: #64888 
Fixes: #64953 
Fixes: #65029 

### Previous Behavior
Slow operation on remote groups or missing remote groups altogether

### New Behavior
Local groups are handled differently from remote ones

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [x] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

File: tests/pytests/functional/utils/user/test__getgrall.py
Comment (whytewolf) [inline tests/pytests/functional/utils/user/test__getgrall.py:1]: can we get some negative testing. such as permission errors, what happens if the group file is not readable by the user the salt-minion is running as. what if there is a duplicate group from groupadd running into MAX_MEMBERS_PER_GROUP.

**Review by whytewolf — COMMENTED** [review]
need a changelog file for 64888

**Review by Ch3LL — DISMISSED** [review]
looks good, just looks like there are test failures related to these changes.

**Comment by nicholasmhughes** [comment]
@Ch3LL , I think I need to remove the "permission denied" test. It works on my machine (TM), but something seems to be allowing access to the file despite removing permissions or the pytest skip mark isn't being respected. Are you okay with me pulling it? It's unlikely to fire in the wild anyway...

**Comment by whytewolf** [comment]
> @Ch3LL , I think I need to remove the "permission denied" test. It works on my machine (TM), but something seems to be allowing access to the file despite removing permissions or the pytest skip mark isn't being respected. Are you okay with me pulling it? It's unlikely to fire in the wild anyway...

I'm ok with the permission test being removed. it is a very rare case it might happen.

**Review by whytewolf — APPROVED** [review]

**Review by garethgreenaway — APPROVED** [review]

**Pull request merged**